### PR TITLE
Allow ProjectSystemCache to return projects by unambiguous short names

### PR DESCRIPTION
**Pull request by joelverhagen**
Fix https://github.com/NuGet/Home/issues/3732

/cc @alpaix @emgarten @rrelyea

### CODE REVIEW

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs [hunk 1]
```diff
@@ -318,26 +318,31 @@ public bool TryGetProjectNameByShortName(string name, out ProjectNames projectNa
...
+                return TryGetProjectNameByShortNameWithoutLock(name, out projectNames);
             }
             finally
             {
                 _readerWriterLock.ExitReadLock();
             }
+        }
+
+        private bool TryGetProjectNameByShortNameWithoutLock(string name, out ProjectNames projectNames)
+        {
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs:334]: `Debug.Assert(_readerWriterLock.IsReadLockHeld);`

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs [hunk 2]
```diff
@@ -318,26 +318,31 @@ public bool TryGetProjectNameByShortName(string name, out ProjectNames projectNa
...
+            projectNames = null;
+
+            HashSet<ProjectNames> values;
+            if (_shortNameCache.TryGetValue(name, out values))
+            {
+                // Get the item at the front of the queue
```
Comment (emgarten) [inline src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs]: This comment could be improved, from the tests it looks like this only returns if there are no ambiguous conflicts

Looks like it was this way before also

**Review by emgarten — APPROVED** [review]

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs [hunk 3]
```diff
@@ -432,7 +437,8 @@ private bool TryGetCacheEntry(string secondaryKey, out CacheEntry cacheEntry)
             try
             {
                 ProjectNames primaryKey;
-                if (_projectNamesCache.TryGetValue(secondaryKey, out primaryKey))
+                if (TryGetProjectNameByShortNameWithoutLock(secondaryKey, out primaryKey) ||
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs]: yes, original order was reverse

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs (hunk 2 above)
Comment (joelverhagen) [reply src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs]: Fixed.

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs]: Fixed.

**Pull request merged**